BLUEBUTTON-949: RFC process for Beneficiary FHIR Server (and related systems)

karlmdavis opened the pull request:
First draft, ready for review.

https://jira.cms.gov/browse/BLUEBUTTON-949

karlmdavis commented:
Just requested review (and feedback and suggestions) from a large-but-still-limited group of folks.

Depending on how that goes, in a day or two, I'll tag the whole Blue Button ecosystem (Auth/frontend, BCDA, DPC, MCT, etc.). This process, as written, only applies to the Blue Button Data/backend systems, but I'd like to see all OEDA DASG programs adopt it eventually as I lowkey hate our current Confluence-dependent review processes.

karlmdavis commented:
@StewGoin Thanks for the thoughtful review!

> FOUO / SCMI
Can you de-acronym-ify that for me? Not sure what you're trying to say?

> The major catch for me is I don't know that we can meaningfully discuss some of the potential RFCs in open public
Can you find any examples from our friends listed in the **Prior Art** section that would need to be non-public?

> ... by only allowing vagaries imo.
What do you mean?

StewGoin commented:
> @StewGoin Thanks for the thoughtful review!
> 
> > FOUO / SCMI
> 
> Can you de-acronym-ify that for me? Not sure what you're trying to say?
FOUO == For Official Use Only
SCMI == System Configuration Management Information
> 
> > The major catch for me is I don't know that we can meaningfully discuss some of the potential RFCs in open public
> 
> Can you find any examples from our friends listed in the **Prior Art** section that would need to be non-public?
> 
Any of the SIAs (they are marked FOUO in the original forms from the Library, the wikified version doesn't generally *mark* form data. Any of the tech specs that include diagrams / descriptions of security controls, environment configuration, etc... Along with any that reference pre-decisional / pre-public items (which is most of them from what I can see). Whenever they contain non-public information they can't be effectively published through this mechanism (particularly before official announcement of things).

> > ... by only allowing vagaries imo.
> 
> What do you mean?

The way around unintentionally releasing non-public info is to often be vague, but vagaries would seem to undermine a lot of the purpose here.

karlmdavis commented:
@StewGoin Might you spend some time putting together some edits on Monday that'd address your concerns?

As for me, I think that the ARS and open source policy conflicts might be at least somewhat deliberate: "that's non-public; you can't make it public" becomes a counterproductive, self-reinforcing loop.

Rather than that framework, I propose the following for evaluating risk and determining what needs to be encrypted:

1. Are there other public systems we respect that are intentionally posting this level of detail?
2. If not, do we have a _damned good reason_ for posting it ourselves?
3. Run _any & all_ risk management concerns by your ISSO, if they object _at all_, encrypt things.

That said, I can't see anything posted in the Blue Button FE's architectural decisions that isn't cool to share publicly, or envision any backend architectural RFCs that also wouldn't be cool. But mebbe I'm just a hippy.

StewGoin commented:
@karlmdavis I'll start down a few compensating controls type items, but fundamentally this kind of approach will need a few levels of business ownership support to enable us to try and do our own thing when it comes to risk management vs. the existing approach and its policies.

karlmdavis commented:
Here's a thought that's giving me a bit of clarity on this:

> Imagine you "bought" this as an open source solution "off the shelf" and deployed it per the project's recommendations and tooling.
>
> What would be public information? **That's** what should be public for our projects.

This is what the open source mandate is trying to get at. Otherwise, the private-or-public decisions become very circular (in a negative feedback loop) kind of way.

StewGoin commented:
Based on our internal business/project discussions -- it seems like we've got some next steps to develop the right processes, risk mitigations, and guidance on "how" and "what" etc..

Beyond *those* more governance type things, I think we need more team interaction on the RFC proposal itself by everyone else 😄

karlmdavis commented:
> I think we need more team interaction on the RFC proposal itself by everyone else 😄

@StewGoin Agreed.

Hey, @dho937, @whytheplatypus, @RickHawesUSDS, @rnagle: What would your Goldilocks-style "just right" technical change proposal template contain? Personally, this one feels a bit too long to me, but I don't have any examples of ones that are better.

RickHawesUSDS commented:
@karlmdavis The [Swift evolution process](https://github.com/apple/swift-evolution) is working pretty well. It has an overall direction, an openness to new ideas and contributors, active discussion forums, and clear roles for decisions making.

karlmdavis commented:
> @karlmdavis The [Swift evolution process](https://github.com/apple/swift-evolution) is working pretty well. It has an overall direction, an openness to new ideas and contributors, active discussion forums, and clear roles for decision making.

@RickHawesUSDS Thanks! I noticed a couple of major differences between their process and Rust's:

1. It was structured much better.
    * I liked their organization much better, so I just pushed a couple commits that move things more in that direction.
2. It's much shorter.
    * Having now filled out our template (mostly based on the Rust one, so far) a couple times, I found myself okay with the length. More importantly, I found the prompts in there to be largely useful.
    * Nonetheless, I added a mention at the top of the template to not just fill things out for the sake of filling things out. I also removed some of the less-applicable prompts I'd copied over, and edited many of the others.

Thoughts?

karlmdavis commented:
Worth noting the RDS Aurora RFC I drafted using this process here: https://github.com/CMSgov/bluebutton-data-model/pull/63

In general, I found the template pretty reasonable there! It helped guide me to exploring and documenting useful things.

karlmdavis commented:
Based on the feedback received here and also offline, I think this is probably "good enough for now" -- not perfect, but a good place to start.

Accordingly, I'm making an official "motion for final comment," with a disposition to *merge* this PR.

If you agree, please give this comment a 👍. If you disagree, give it a 👎. If you'd like to officially abstain, give it a 🦆.

nickrobison-usds commented on rfcs/0001-beneficiary-fhir-server-rfc-process-v1.md:
```diff
@@ -0,0 +1,168 @@
+* [./0000-template.md](Template/Starter RFC)
+    * Intended to be copy-pasted for each new proposed RFC.
+
+### Proposed Solution: Compensatory Steps
```
This section is pretty unclear to me. What are 'compensatory steps'? Are we simply outlining why and how this is different from just using a Confluence doc?

karlmdavis replied:
I believe the intent was to detail how our _behavior_ should differ from a more private process (i.e. one in Confluence).

Regardless, this guidance isn't fully-baked yet and it makes more sense for it to live in Confluence, so I've moved it there in my most recent update.

nickrobison-usds commented on rfcs/0001-beneficiary-fhir-server-rfc-process-v1.md:
```diff
@@ -0,0 +1,168 @@
+What do we have to take care in doing, and perhaps do differently than if this were conducted on Confluence?
+
+* No information we wouldn't be comfortable having included as part of a news article
+    (say one by Brian Krebs) should be useed unencrypted --
```
Typo in used.

nickrobison-usds commented on rfcs/0001-beneficiary-fhir-server-rfc-process-v1.md:
```diff
@@ -0,0 +1,168 @@
+* No information we wouldn't be comfortable having included as part of a news article
+    (say one by Brian Krebs) should be useed unencrypted --
+    public data needs to be truly public in nature, not just obscurely comfortable.
+* No information that rises to FIPS 199 moderate levels without an internal agreement to reclassify it as public.
```
These data disclosure rules should probably be in their own sub-section.

karlmdavis replied:
Moved to Confluence, per previous comment.

nickrobison-usds commented on rfcs/README.md:
```diff
@@ -0,0 +1,108 @@
+There'a no harm in throwing up an RFC as a "trial balloon" but a much faster way to get feedback on an idea first... is to _talk about it_ with others.
+Accordingly, you ought to have these conversations (e.g. in Slack) first, **before** publishing your RFC.
+
+## The RFC Process
```
What is the RFC process?

karlmdavis replied:
This was pretty confusing, I think, but the process was detailed in the early step in the "Approval" section.

I restructured these sections to be more clear. Do they make sense to you now?

jzulim commented:
This process and supporting template look great. I suspect this process and template will morph as we begin to use it more but as a first run and with excellent feedback from Steve and Rich, this feels like a fantastic baseline and I support it.

karlmdavis commented:
The following folks appear to have voted to approve this for **merge**: @jzulim, @StewGoin, @RickHawesUSDS, @cthayes, @karlmdavis, and @dho937.

Approved!

karlmdavis commented:
Also: many thanks to everyone who provided feedback: this is a much better process as a result!

❤️

ibh1127 commented:
👍
